File: 0021-merge-two-sorted-lists/0021-merge-two-sorted-lists.dart
/**
 * Definition for singly-linked list.
 * class ListNode {
 *   int val;
 *   ListNode? next;
 *   ListNode([this.val = 0, this.next]);
 * }
 */
class Solution {
  ListNode? mergeTwoLists(ListNode? list1, ListNode? list2) {
    if (list1 == null && list2 == null) {
        return null;
    } else if (list1 == null && list2 != null) {
        return list2;
    } else if (list1 != null && list2 == null) {
        return list1;
    }
    var maintain = ListNode(-1);
    dynamic answer = maintain;
    while (list1 != null && list2 != null) {
        if (list1.val >= list2.val) {
            answer.next = list2;
            list2 = list2.next;
        } else {
            answer.next = list1;
            list1 = list1.next;
        }
        answer = answer.next;
    }
    if (list1 == null && list2 != null) {
        answer.next = list2;
    } else if (list1 != null && list2 == null) {
        answer.next = list1;
    }

    return maintain.next;
  }
}

File: 0021-merge-two-sorted-lists/0021-merge-two-sorted-lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeTwoLists(
        self, list1: Optional[ListNode], list2: Optional[ListNode]
    ) -> Optional[ListNode]:
        maintain = ListNode(-1)
        answer = maintain
        if not list1 and not list2:
            return None
        if not list1 and list2:
            return list2
        if list1 and not list2:
            return list1
        if list1 and list2:
            while list1 and list2:
                if list1.val <= list2.val:
                    answer.next = list1
                    list1 = list1.next
                else:
                    answer.next = list2
                    list2 = list2.next
                answer = answer.next
            answer.next = self.mergeTwoLists(list1, list2)
        return maintain.next
